README.md

```diff
@@ -6,7 +6,498 @@
+
+  Enabled by importing `faststreams/chronos_adapters`. <br />
+  It can represent any Chronos `Transport` as an input stream.
+
+* `asyncSocketOutput` (async)
+
+  Enabled by importing `faststreams/std_adapters`. <br />
+  Allows using Nim's standard library `AsyncSocket` type as an output stream.
+
+You can extend the library with new `OutputStream` types without modifying it.
+Please see the inline code documentation of `OutputStreamVTable` for more details.
+
+All of the above APIs are possible constructors for creating an `OutputStream`.
+The stream instances will manage their resources through destructors, but you
+might want to `close` them explicitly in async context or when you need to
+handle the possible errors from the closing operation.
+
+Here is an example usage:
+
+```nim
+import
+  faststreams/outputs
+
+type
+  ABC = object
+    a: int
+    b: char
+    c: string
+
+var stream = memoryOutput()
+stream.writeNimRepr(ABC(a: 1, b: 'b', c: "str"))
+var repr = stream.getOutput(string)
+```
+
+The `writeNimRepr` in the above example is not part of the library, but
+let's see how it can be implemented:
+
+```nim
+import
+  typetraits, faststreams/outputs
+
+proc writeNimRepr*(stream: OutputStream, str: string) =
+  stream.write '"'
+
+  for c in str:
+    if c == '"':
+      stream.write ['\'', '"']
+    else:
+      stream.write c
+
+  stream.write '"'
+
+proc writeNimRepr*(stream: OutputStream, x: char) =
+  stream.write ['\'', x, '\'']
+
+proc writeNimRepr*(stream: OutputStream, x: int) =
+  stream.write $x # Making this more optimal has been left
+                  # as an exercise for the reader
+
+proc writeNimRepr*[T](stream: OutputStream, obj: T) =
+  stream.write typetraits.name(T)
+  stream.write '('
+
+  var firstField = true
+  for name, val in fieldPairs(obj):
+    if not firstField:
+      stream.write ", "
+
+    stream.write name
+    stream.write ": "
+    stream.writeNimRepr val
+
+    firstField = false
+
+  stream.write ')'
+```
+
+When the stream is created, its output buffers will be initialized with a
+single page of `pageSize` bytes (specified at stream creation). Calls to
+`write` will just populate this page until it becomes full and only then
+it would be sent to the output device.
+
+As the example demonstrates, a `memoryOutput` will continue buffering
+pages until they can be finally concatenated and returned in `stream.getOutput`.
+If the output fits within a single page, it will be efficiently moved to
+the `getOutput` result. When the output size is known upfront you can ensure
+that this optimization is used by calling `stream.ensureRunway` before any
+writes, but please note that the library is free to ignore this hint in async
+context or if a maximum memory usage policy is specified.
+
+In a non-memory stream, any writes larger than a page or issued through the
+`writeNow` API will be sent to the output device immediately.
+
+Please note that even in async context, `write` will complete immediately.
```

Comment:
I wonder about this default: ie incomplete writes for files is such a common mistake and it's defaults like this in the API that cause it - on networks, it's even more common that writes fail. `tryWrite` perhaps would remind people of this fact, because I'd agree in general that a function that doesn't wait/block often is a good tradeoff

Reply:
basically, this is where data corruption happens, causing users to lose their work because developers don't consider the hidden delayed action - it's satisfying to the developer that feels that the library is easy to use because they don't have to bother with the details, but less satisfying to the user

Reply:
furthering this point - `when await s.readable` is a natural and "forced" collection point where error checking is done for reading the stream whereas for writes it remains "optional" - what would be a way to introduce a similar "forced" checkpoint for writes? if it's not "naturally" part of the API, it won't get called (just like `close` de-facto doesn't get caled with users paying the price for the poor API)

Reply:
another point: ignoring backpressure and making it "optional" will not fly - it's not an aspect that you _can_ ever ignore - you have to account for it in every application that does async writing or your application is broken _in the cases where performance actually matters, ie when writing a lot of data_, so there has to be a point in the faststreams API that forces the user to deal with it. `write` is the traditional place in `unix` etc - either it's synchronous or it returns `EAGAIN` - if faststreams is to offer an async mode, it necessarily needs to have a forced synchronization point that the caller of the code cannot ignore.

it really doesn't matter that the library is fast when writing 1 byte if it unexpectedly breaks with OOM when it writes 1gb.

Reply:
practical example: the USB stick that you have to "eject" before removing safely - in the 00's this was thought of as a great trade off  because it was _so fast_ but the world leaned better and now doesn't force users to go through this silliness - instead, copy operations visibly don't terminate until the writing is physically done.

Reply:
I consider this a bit experimental myself and might agree to change the default. The point is that the library offers:

* `write` (buffers data and puts a future in `lastWriteFut` when you can check it if you wish) 
* `writeAndWait` (the standard write with back-pressure)
* `writeNow` (this a non-buffered write) 

It's just a matter of selecting which of the two should be promoted as the default way. If we make `writeAndWait` the default, what would be a good name for `write`?